### PR TITLE
Trying to fix react 15

### PR DESCRIPTION
**Pull request by theKashey**
Bulk changes to fix all the things.
It still not always working :(

### CODE REVIEW

**Comment by neoziro** [comment]
I think your way to forceUpdate is not correct. We should maybe rely on react-force-update package.

File: packages/react-hot-loader/src/AppContainer.dev.js [hunk 1]
```diff
@@ -43,7 +43,9 @@ class AppContainer extends React.Component {
       })
 
       // perform sandboxed render to find similarities between new and old code
+      this.renderAnotherChildren = nextProps.children;
```
Comment (neoziro) [inline packages/react-hot-loader/src/AppContainer.dev.js]: What does it fix?

File: packages/react-hot-loader/src/reconciler/hotReplacementRender.js [hunk 2]
```diff
@@ -106,45 +111,49 @@ const mergeInject = (a, b) => {
 }
 
 const hotReplacementRender = (instance, stack) => {
-  const flow = asArray(render(instance))
+  const flow = filterNullArray(asArray(render(instance)))
```
Comment (neoziro) [inline packages/react-hot-loader/src/reconciler/hotReplacementRender.js]: This is not perf, we iterate two times. I prefer to check child after.

File: packages/react-hot-loader/src/reconciler/hotReplacementRender.js [hunk 3]
```diff
@@ -106,45 +111,49 @@ const mergeInject = (a, b) => {
...
+    if (typeof child !== 'object' || !child.type || !stackChild.instance) {
       return
     }
+
     if (typeof child.type !== 'function') {
       next(
         // move types from render to the instances of hydrated tree
         mergeInject(
-          child.props ? child.props.children : child.children,
+          filterNullArray(asArray(child.props ? child.props.children : child.children)),
```
Comment (neoziro) [inline packages/react-hot-loader/src/reconciler/hotReplacementRender.js]: This is not the fix.

File: packages/react-hot-loader/src/reconciler/reactUtils.js [hunk 4]
```diff
@@ -12,5 +14,7 @@ export const updateInstance = instance => {
     instance.forceUpdate()
   } else if (updater && typeof updater.enqueueForceUpdate === 'function') {
     updater.enqueueForceUpdate(instance)
+  } else {
+    instance._pendingForceUpdate = true
```
Comment (neoziro) [inline packages/react-hot-loader/src/reconciler/reactUtils.js]: Useless

File: packages/react-hot-loader/src/reconciler/reactUtils.js [hunk 5]
```diff
@@ -1,7 +1,9 @@
 /* eslint-disable no-underscore-dangle */
 
 export const getReactInstance = instance =>
-  instance._reactInternalFiber || instance._reactInternalInstance
+  instance._reactInternalFiber ||
+  instance._reactInternalInstance ||
+  instance._instance
```
Comment (neoziro) [inline packages/react-hot-loader/src/reconciler/reactUtils.js]: OK

**Comment by neoziro** [comment]
Instead of reactHydrate strategy why not walking the React tree, swap what is swappable, mark as pending update. Then walk again and update it?

**Comment by theKashey** [comment]
@neoziro - I've fix close to everything. Tested on pack of local projects - everything is awesome.
Changes:
 1. patch.dev proxies _any_ components. (not only with regenerate method, as before)
 2. reconciler also monkeypatching props during test render. Thus allow to broadcast `not-registered` changes via props (render/children)
 3. Before render - call componentWillUpdate. This is the only way to drop Redux's renderedComponent cache
 4. (Stand-in) After render try to wrap result with proxy. This is the only way to wrap with proxy cached rendered component from (yet again) Redux.

Also - I think few moments could be simplified. Did not test moment about registering the component in `hot` - may be it is not necessary (but it could).

**Pull request merged**